File: Plot_Maker.m
clear
clc

% Read the CSV file into a table
data = readtable('optimization_data_x1n30.csv');

Stress_vector = data.MaxStress;
Mass_W_P_vector = data.MassWithoutPenalty;
Mass_F_A_vector = data.MassFromAbaqus;
Objective_vector = data.Objective;

x = (1:31);

% Ploting Max Stress vs. Iteration
figure(1)
plot(x, Stress_vector)
xlim([1 31])
title('Max Stress vs. Iteration')
xlabel('Iterations')
ylabel('Max Stress (MPa)')
 
% Ploting Mass Without Penalty vs. Iteration
figure(2)
plot(x, Mass_W_P_vector)
xlim([1 31])
title('Mass Without Penalty vs. Iteration')
xlabel('Iterations')
ylabel('Mass Without Penalty (tonnes)')
 
% Ploting Mass (Abaqus) vs. Iteration
figure(3)
plot(x, Mass_F_A_vector)
xlim([1 31])
title('Mass (Abaqus) vs. Iteration')
xlabel('Iterations')
ylabel('Mass (Abaqus) (tonnes)')
 
% Ploting Objective Value vs. Iteration
figure(4)
plot(x, Objective_vector)
xlim([1 31])
title('Objective Value vs. Iteration')
xlabel('Iterations')
ylabel('Objective Value')

% Making the 3 way plot for B.O.
data2 = readtable('optimization_data_x2n30');
data3 = readtable('optimization_data_x3n30');

Objective_vector_2 = data2.Objective;
Objective_vector_3 = data3.Objective;

Mass_F_A_vector_2 = data2.MassFromAbaqus;
Mass_F_A_vector_3 = data3.MassFromAbaqus;

Mass_W_P_vector_2 = data2.MassWithoutPenalty;
Mass_W_P_vector_3 = data3.MassWithoutPenalty;

Stress_vector_2 = data2.MaxStress;
Stress_vector_3 = data3.MaxStress;

x_2 = (1:30);
x_3 = (1:30);

figure (5)
plot(x,Objective_vector)
hold on
plot(x_2,Objective_vector_2)
plot(x_3,Objective_vector_3)
xlim([1 32])
title('Objective Value For Three Separate Runs')
xlabel('Iterations')
ylabel('Objective Value')
hold off

figure (6)
plot(x,Mass_F_A_vector)
hold on
plot(x_2,Mass_F_A_vector_2)
plot(x_3,Mass_F_A_vector_3)
xlim([1 32])
title('Mass (Abaqus) For Three Separate Runs')
xlabel('Iterations')
ylabel('Mass (Abaqus) (tonnes)')
hold off

figure (7)
plot(x,Mass_W_P_vector)
hold on
plot(x_2,Mass_W_P_vector_2)
plot(x_3,Mass_W_P_vector_3,'MarkerFaceColor',[0.9290 0.6940 0.1250]);
xlim([0 30])
title('Mass Without Penalty For Three Separate Runs')
xlabel('Iterations')
ylabel('Mass Without Penalty (tonnes)')
plot(1,3.69,'o','MarkerEdgeColor',[0.4940 0.1840 0.5560], 'MarkerSize', 10)
plot(1,2.79,'o','MarkerEdgeColor',[0.4660 0.6740 0.1880], 'MarkerSize', 10)
plot(1,2.43,'o','MarkerEdgeColor',[0.3010 0.7450 0.9330], 'MarkerSize', 10)
plot(7,4.07,'*','MarkerEdgeColor',[0.4940 0.1840 0.5560], 'MarkerSize', 10)
plot(11,4.76,'*','MarkerEdgeColor',[0.4660 0.6740 0.1880], 'MarkerSize', 10)
plot(10,3.26,'*','MarkerEdgeColor',[0.3010 0.7450 0.9330], 'MarkerSize', 10)
L = legend('Run 1','Run 2','Run 3','(x0,y0)=(1,3.69)','(x0,y0)=(1,2.79)','(x0,y0)=(1,2.43)','(x*,y*)=(7,4.07)','(x*,y*)=(11,4.76)','(x*,y*)=(10,3.26)');
L.AutoUpdate = 'off';
scatter(x, Mass_W_P_vector, 10, 'k', 'filled');
scatter(x_2,Mass_W_P_vector_2, 10,'k', 'filled')
scatter(x_3,Mass_W_P_vector_3, 10,'k', 'filled')
hold off

figure (8)
plot(x,Stress_vector)
hold on
plot(x_2,Stress_vector_2)
plot(x_3,Stress_vector_3)
xlim([1 32])
title('Max Stress For Three Separate Runs')
xlabel('Iterations')
ylabel('Max Stress (MPa)')
hold off

data4 = readtable('random_results.csv');

Mass_F_A_vector_4 = data4.Aba_Mass;

Mass_W_P_vector_4 = data4.Mass;

Stress_vector_4 = data4.MaxStress;

x_4 = (1:32);

figure (9)
plot(x,Mass_W_P_vector)
hold on
plot(x_4,Mass_W_P_vector_4)
xlim([0 30])
plot(1,3.69,'ko', 'MarkerSize', 10)
plot(7,4.07,'*','MarkerEdgeColor',[0 0.4470 0.7410], 'MarkerSize', 10)
plot(23,4.40,'*','MarkerEdgeColor',[0.8500 0.3250 0.0980], 'MarkerSize', 10)
title('Bayesian Optimization (B.O.) vs. Random Sample (R.S.)')
xlabel('Iterations')
ylabel('Mass Without Penalty (tonnes)')
L = legend('Bayesian Optimization','Random Sample','(x0,y0)=(1,3.69)','B.O. (x*,y*)=(7,4.07)','R.S. (x*,y*)=(23,4.40)');
L.AutoUpdate = 'off';
scatter(x, Mass_W_P_vector, 10, 'k', 'filled');
scatter(x_4,Mass_W_P_vector_4, 10,'k', 'filled')

File: Plot_Maker.py
# Set up Python 
import math
import csv
import subprocess
import re
import os
from skopt import gp_minimize
from skopt.space import Real
from skopt.utils import use_named_args
import matplotlib.pyplot as plt
import pandas as pd

# Load data from three CSV files
data1 = pd.read_csv("optimization_data_60_x1.csv")  # First dataset
data2 = pd.read_csv("optimization_data_60_x2.csv")  # Second dataset
data3 = pd.read_csv("optimization_data_60_x3.csv")  # Third dataset

# Extract iteration number (row index) and objective values
iterations1 = range(0, len(data1))  # Start from 0
iterations2 = range(0, len(data2))  # Start from 0
iterations3 = range(0, len(data3))  # Start from 0

objective_values1 = data1["Objective"].astype(float)
objective_values2 = data2["Objective"].astype(float)
objective_values3 = data3["Objective"].astype(float)

max_stress1 = data1["MaxStress"].astype(float)
max_stress2 = data2["MaxStress"].astype(float)
max_stress3 = data3["MaxStress"].astype(float)

mass_without_penalty1 = data1["MassWithoutPenalty"].astype(float)
mass_without_penalty2 = data2["MassWithoutPenalty"].astype(float)
mass_without_penalty3 = data3["MassWithoutPenalty"].astype(float)

# Identify the starting and optimal design for all datasets
initial_index1 = 0  # First iteration
optimal_index1 = objective_values1.idxmin()  # Index of minimum objective value

initial_index2 = 0  # First iteration
optimal_index2 = objective_values2.idxmin()  # Index of minimum objective value

initial_index3 = 0  # First iteration
optimal_index3 = objective_values3.idxmin()  # Index of minimum objective value

# Plot 1: Optimization Progression
plt.figure(figsize=(8, 5))
plt.plot(iterations1, objective_values1, marker='o', linestyle='-', markersize=5, label="Number of Calls 60")
plt.plot(iterations2, objective_values2, marker='s', linestyle='--', markersize=5, label="Number of Calls 60")
plt.plot(iterations3, objective_values3, marker='^', linestyle='-.', markersize=5, label="Number of Calls 50")

plt.scatter(iterations1[initial_index1], objective_values1[initial_index1], color='red', s=100, label="Initial Design 1", zorder=3)
plt.scatter(iterations2[initial_index2], objective_values2[initial_index2], color='purple', s=100, label="Initial Design 2", zorder=3)
plt.scatter(iterations3[initial_index3], objective_values3[initial_index3], color='orange', s=100, label="Initial Design 3", zorder=3)

plt.scatter(iterations1[optimal_index1], objective_values1[optimal_index1], color='green', s=100, label="Optimal Design 1", zorder=3)
plt.scatter(iterations2[optimal_index2], objective_values2[optimal_index2], color='blue', s=100, label="Optimal Design 2", zorder=3)
plt.scatter(iterations3[optimal_index3], objective_values3[optimal_index3], color='brown', s=100, label="Optimal Design 3", zorder=3)

plt.xlabel("Iteration")
plt.ylabel("Objective (Mass + Penalty)")
plt.title("Optimization Progress: Objective vs. Iteration")
plt.legend()
plt.grid(True)
plt.show()

# Plot 2: Max Stress vs Iteration
plt.figure(figsize=(8, 5))
plt.plot(iterations1, max_stress1, marker='o', linestyle='-', markersize=5, label="Number of Calls 60")
plt.plot(iterations2, max_stress2, marker='s', linestyle='--', markersize=5, label="Number of Calls 60")
plt.plot(iterations3, max_stress3, marker='^', linestyle='-.', markersize=5, label="Number of Calls 50")

plt.scatter(iterations1[initial_index1], max_stress1[initial_index1], color='red', s=100, label="Initial Design 1", zorder=3)
plt.scatter(iterations2[initial_index2], max_stress2[initial_index2], color='purple', s=100, label="Initial Design 2", zorder=3)
plt.scatter(iterations3[initial_index3], max_stress3[initial_index3], color='orange', s=100, label="Initial Design 3", zorder=3)

plt.scatter(iterations1[optimal_index1], max_stress1[optimal_index1], color='green', s=100, label="Optimal Design 1", zorder=3)
plt.scatter(iterations2[optimal_index2], max_stress2[optimal_index2], color='blue', s=100, label="Optimal Design 2", zorder=3)
plt.scatter(iterations3[optimal_index3], max_stress3[optimal_index3], color='brown', s=100, label="Optimal Design 3", zorder=3)

plt.xlabel("Iteration")
plt.ylabel("Max Abaqus Stress (MPa)")
plt.title("Max Stress vs. Iteration")
plt.legend()
plt.grid(True)
plt.show()

# Plot 3: Mass Without Penalty vs Iteration
plt.figure(figsize=(8, 5))
plt.plot(iterations1, mass_without_penalty1, marker='o', linestyle='-', markersize=5, label="Number of Calls 60")
plt.plot(iterations2, mass_without_penalty2, marker='s', linestyle='--', markersize=5, label="Number of Calls 60")
plt.plot(iterations3, mass_without_penalty3, marker='^', linestyle='-.', markersize=5, label="Number of Calls 50")

plt.scatter(iterations1[initial_index1], mass_without_penalty1[initial_index1], color='red', s=100, label="Initial Design 1", zorder=3)
plt.scatter(iterations2[initial_index2], mass_without_penalty2[initial_index2], color='purple', s=100, label="Initial Design 2", zorder=3)
plt.scatter(iterations3[initial_index3], mass_without_penalty3[initial_index3], color='orange', s=100, label="Initial Design 3", zorder=3)

plt.scatter(iterations1[optimal_index1], mass_without_penalty1[optimal_index1], color='green', s=100, label="Optimal Design 1", zorder=3)
plt.scatter(iterations2[optimal_index2], mass_without_penalty2[optimal_index2], color='blue', s=100, label="Optimal Design 2", zorder=3)
plt.scatter(iterations3[optimal_index3], mass_without_penalty3[optimal_index3], color='brown', s=100, label="Optimal Design 3", zorder=3)

plt.xlabel("Iteration")
plt.ylabel("Mass Without Penalty (tonnes)")
plt.title("Mass Without Penalty vs. Iteration")
plt.legend()
plt.grid(True)
plt.show()
